cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterCreator.java

```diff
@@ -0,0 +1,267 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utility class to help to create the KafkaCluster model
+ */
+public class KafkaClusterCreator {
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaClusterCreator.class.getName());
+
+    // Settings
+    private final Reconciliation reconciliation;
+    private final boolean useKRaftFGEnabled;
+    private final KafkaVersion.Lookup versions;
+
+    // Operators and other tools
+    private final Vertx vertx;
+    private final AdminClientProvider adminClientProvider;
+    private final SecretOperator secretOperator;
+    private final SharedEnvironmentProvider sharedEnvironmentProvider;
+    private final PreventBrokerScaleDownCheck brokerScaleDownOperations;
+
+    // State
+    private boolean scaleDownCheckFailed = false;
+
+    /**
+     * Constructor
+     *
+     * @param vertx             Vert.x instance
+     * @param reconciliation    Reconciliation marker
+     * @param config            Cluster Operator configuration
+     * @param supplier          Resource Operators supplier
+     */
+    public KafkaClusterCreator(
+            Vertx vertx,
+            Reconciliation reconciliation,
+            ClusterOperatorConfig config,
+            ResourceOperatorSupplier supplier
+    ) {
+        this.reconciliation = reconciliation;
+        this.versions = config.versions();
+        this.useKRaftFGEnabled = config.featureGates().useKRaftEnabled();
+
+        this.vertx = vertx;
+        this.adminClientProvider = supplier.adminClientProvider;
+        this.secretOperator = supplier.secretOperations;
+        this.sharedEnvironmentProvider = supplier.sharedEnvironmentProvider;
+        this.brokerScaleDownOperations = supplier.brokerScaleDownOperations;
+    }
+
+    /**
+     * Prepares the Kafka Cluster model instance. It checks if any scale-down is happening and whether such scale-down
+     * can be done. If it discovers any problems, it will try to fix them and create a fixed Kafka Cluster model
+     * instance if the {@code tryToFixProblems} fag is set to true. It throws an exception otherwise.
+     *
+     * @param kafkaCr           Kafka custom resource
+     * @param nodePools         List with Kafka Node Pool resources
+     * @param oldStorage        Old storage configuration
+     * @param currentPods       Existing Kafka pods
+     * @param versionChange     Version Change object describing any possible upgrades / downgrades
+     * @param tryToFixProblems  Flag indicating whether recoverable configuration issues should be fixed or not
+     *
+     * @return  New Kafka Cluster instance
+     */
+    public Future<KafkaCluster> prepareKafkaCluster(
+            Kafka kafkaCr,
+            List<KafkaNodePool> nodePools,
+            Map<String, Storage> oldStorage,
+            Map<String, List<String>> currentPods,
+            KafkaVersionChange versionChange,
+            boolean tryToFixProblems
```

Comment:
why this flag? You envisage scenarios in the code where we don't want to fix problems? (Asking because you are already using this method but of course passing true here).

Reply:
Because it is called recursively -> in the first call, we try to fix things. In the second, if there are any errors we failed to fix, it will just the the error and fail the reconciliation. This flag is basically used to indicate which behavior to activate.